metrics/ci_integrations/lib/source/github_actions/adapter/github_actions_source_client_adapter.dart

```diff
@@ -0,0 +1,326 @@
+import 'package:metrics_core/metrics_core.dart';
+
+/// An adapter for [GithubActionsClient] to implement the [SourceClient]
+/// interface.
+class GithubActionsSourceClientAdapter implements SourceClient {
+  /// A fetch limit for the paginated requests.
+  static const perPageLimit = 25;
+
+  /// A [GithubActionsClient] instance to perform API calls.
+  final GithubActionsClient githubActionsClient;
+
+  /// Creates a new instance of the [GithubActionsSourceClientAdapter].
+  ///
+  /// Throws an [ArgumentError] if the given [githubActionsClient] is `null`.
+  GithubActionsSourceClientAdapter(this.githubActionsClient) {
+    ArgumentError.checkNotNull(githubActionsClient, 'githubActionsClient');
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuilds(String workflowIdentifier) async {
+    final runsPage = await _fetchRunsPage(
+      workflowIdentifier: workflowIdentifier,
+      status: RunStatus.completed,
+      perPage: perPageLimit,
+      page: 1,
+    );
+
+    final runs = runsPage.values;
+
+    return _mapRunsToBuildData(
+      runs: runs,
+      workflowIdentifier: workflowIdentifier,
+    );
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuildsAfter(
+    String workflowIdentifier,
+    BuildData build,
+  ) async {
+    final initialRunsPage = await _fetchRunsPage(
+      workflowIdentifier: workflowIdentifier,
+      status: RunStatus.completed,
+      perPage: 1,
+      page: 1,
+    );
+
+    final lastRunNumber = initialRunsPage.values.first.number;
+    final lastBuildNumber = build.buildNumber;
+
+    if (lastRunNumber <= lastBuildNumber) return [];
+
+    final runsAfter = await _fetchRunsAfter(
+      workflowIdentifier: workflowIdentifier,
+      lastBuildNumber: lastBuildNumber,
+    );
+
+    return _mapRunsToBuildData(
+      runs: runsAfter,
+      workflowIdentifier: workflowIdentifier,
+    );
+  }
+
+  /// Fetches a list of [WorkflowRun]s that have the [WorkflowRun.runNumber]
+  /// greater than the given [lastBuildNumber].
+  ///
+  /// A [workflowIdentifier] is either a workflow id or a name of the file
+  /// that defines the workflow.
+  Future<List<WorkflowRun>> _fetchRunsAfter({
+    String workflowIdentifier,
+    int lastBuildNumber,
+  }) async {
+    WorkflowRunsPage page = await _fetchRunsPage(
+      workflowIdentifier: workflowIdentifier,
+      status: RunStatus.completed,
+      perPage: perPageLimit,
+      page: 1,
+    );
+
+    final List<WorkflowRun> result = [];
+
+    while (page.hasNextPage) {
+      final runs = page.values;
+
+      for (final run in runs) {
+        if (run.number > lastBuildNumber) {
+          result.add(run);
+        }
+        if (run.number == lastBuildNumber) return result;
+      }
+
+      page = await _fetchNextRunsPage(page);
+    }
+
+    return result;
+  }
+
+  /// Fetches the [WorkflowRunDuration] of the given [run].
+  Future<WorkflowRunDuration> _fetchRunDuration(WorkflowRun run) async {
```

Comment:
Let's return a `Duration` instance here. 
```suggestion
  Future<Duration> _fetchRunDuration(WorkflowRun run) async {
```